src/Compilers/CSharp/Portable/Symbols/Source/SourceNamespaceSymbol.cs

```diff
@@ -372,13 +372,35 @@ private static void CheckMembers(NamespaceSymbol @namespace, Dictionary<string,
 
                     if ((object)other != null)
                     {
-                        if ((nts as SourceNamedTypeSymbol)?.IsPartial == true && (other as SourceNamedTypeSymbol)?.IsPartial == true)
+                        switch (nts, other)
                         {
-                            diagnostics.Add(ErrorCode.ERR_PartialTypeKindConflict, symbol.Locations.FirstOrNone(), symbol);
+                            case (SourceNamedTypeSymbol left, SourceNamedTypeSymbol right) when isLeftAFileTypeInSeparateSourceTreeFromRight(left, right) || isLeftAFileTypeInSeparateSourceTreeFromRight(right, left):
+                                // no error
+                                break;
+                            case (SourceNamedTypeSymbol { IsPartial: true }, SourceNamedTypeSymbol { IsPartial: true }):
+                                diagnostics.Add(ErrorCode.ERR_PartialTypeKindConflict, symbol.Locations.FirstOrNone(), symbol);
+                                break;
+                            default:
+                                diagnostics.Add(ErrorCode.ERR_DuplicateNameInNS, symbol.Locations.FirstOrNone(), name, @namespace);
+                                break;
                         }
-                        else
+
+                        // prototype(ft): better name? better diagnostic?
+                        static bool isLeftAFileTypeInSeparateSourceTreeFromRight(SourceNamedTypeSymbol left, SourceNamedTypeSymbol right)
```

Comment:
If possible, could we move the local function to the end of the method as to not interrupt the flow of main logic? I'd much appreciate

Reply:
I end up moving to an extension although it's still only used once currently. I think it reads easier.